File: ml-benchmark/synduce/treepaths/maxPathWeight2.f
Inductive BTree = empty Unit | node {Int, BTree, BTree};
Inductive Zipper = top Unit | left {Int, BTree, Zipper} | right {Int, BTree, Zipper};

max = \a: Int. \b: Int. if (< a b) then b else a;

mpath = fix (
  \f: BTree -> Int. \t: BTree.
  match t with
    empty _ -> 0
  | node {a, l, r} -> + a (max (f l) (f r))
  end
);

repr = fix (
  \f: Zipper -> BTree. \z: Zipper.
  match z with
    top _ -> empty unit
  | left {w, tree, zz} ->
      node {w, tree, f zz}
  | right {w, tree, zz} ->
      node {w, f zz, tree}
  end
);

tree_rec = fix (
  \f: BTree -> Compress BTree. \t: BTree.
  match t with
    empty _ -> empty unit
  | node {a, l, r} -> node {a, f l, f r}
  end
);
zip_rec = fix (
  \f: Zipper -> Compress Zipper. \z: Zipper.
  match z with
    top _ -> top unit
  | left {w, tree, zz} -> left {w, tree_rec tree, f zz}
  | right {w, tree, zz} -> right {w, tree_rec tree, f zz}
  end
);

main = \z: Zipper. mpath (repr (zip_rec z));

File: ml-benchmark/synduce/treepaths/maxPathWeight2.ml
type btree = Empty of unit | Node of int * btree * btree
type zipper = Top of unit | Left of int * btree * zipper | Right of int * btree * zipper

let max a b = if a < b then b else a

let rec mpath t =
  match t with
  | Empty _ -> 0
  | Node (a, l, r) -> a + max (mpath l) (mpath r)

let rec repr z =
  match z with
  | Top _ -> Empty ()
  | Left (w, tree, zz) -> Node (w, tree, repr zz)
  | Right (w, tree, zz) -> Node (w, repr zz, tree)

val tree_rec: btree -> btree compress
let rec tree_rec t =
  match t with
  | Empty _ -> Empty ()
  | Node (a, l, r) -> Node (a, tree_rec l, tree_rec r)

val zip_rec: zipper -> zipper compress
let rec zip_rec z =
  match z with
  | Top _ -> Top ()
  | Left (w, tree, zz) -> Left (w, tree_rec tree, zip_rec zz)
  | Right (w, tree, zz) -> Right (w, tree_rec tree, zip_rec zz)

let program z = mpath (repr (zip_rec z))